socks5.py

```diff
@@ -220,57 +315,129 @@ def read_addrport(self, address_type, sockfile):
 
-            if sock1 in r:
-                data = sock1.recv(4096)
+            if csock in r:
+                data = csock.recv(4096)
                 if not data:
                     break
-                sock2.sendall(data)
+                ssock.sendall(data)
+                with traffic_lock:
+                    outbound_traffic += len(data)
 
-            if sock2 in r:
-                data = sock2.recv(4096)
+            if ssock in r:
+                data = ssock.recv(4096)
                 if not data:
                     break
-                sock1.sendall(data)
+                csock.sendall(data)
+                with traffic_lock:
+                    inbound_traffic += len(data)
+
 
     def handle_connect(self, address, port):
         log_tag = '%s:%s -> %s:%s' % (self.client_address + (address, port))
-        try:
-            remote = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            if CONNECT_HOST:
-                remote.bind((CONNECT_HOST, 0))
-            remote.connect((address, port))
-            logging.info('%s: connected', log_tag)
-        except Exception as err:
-            logging.error('%s: connect error %s', log_tag, err)
-            self.send_reply(self.STATUS_ECONNREFUSED)
-            self.server.close_request(self.request)
-            return
 
-        self.send_reply(self.STATUS_SUCCEEDED)
+        if isinstance(address, dict):
+            ipv4 = address.get('ipv4')
+            ipv6 = address.get('ipv6')
+        else:
+            if ':' in address:
+                ipv6 = address
+                ipv4 = None
+            else:
+                ipv4 = address
+                ipv6 = None
+
+        if (ipv6 is not None and CONNECT_HOST_IPV6) and \
+        (ipv4 is not None and CONNECT_HOST_IPV4):
+            with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
+                if ipv6 is not None:
+                    ipv6_future = executor.submit(self.connect_to_address, ipv6, port, socket.AF_INET6, log_tag, timeout=5)
+                if ipv4 is not None:
+                    # Start IPv4 connection after 50ms delay
+                    ipv4_future = executor.submit(self.delayed_connect, ipv4, port, socket.AF_INET, 0.05, log_tag)
+
+                while True:
+                    if ipv6 is not None and ipv6_future.done():
+                        remote = ipv6_future.result()
+                        if remote is not None:
+                            if ipv4 is not None:
+                                ipv4_future.cancel()
+                            break
+                    if ipv4 is not None and ipv4_future.done():
+                        remote = ipv4_future.result()
+                        if remote is not None:
+                            if ipv6 is not None:
+                                ipv6_future.cancel()
+                            break
+                    time.sleep(0.01)  # Avoid busy-waiting
+        else:
+            try:
```

Comment:
nit: need to set `remote` to `None` up here to prevent `local variable 'remote' referenced before assignment` errors when connecting to invalid/non-resolved hosts